block/block.go

```diff
@@ -13,23 +13,27 @@ import (
 // applyBlock applies the block to the store and the abci app.
 // steps: save block -> execute block with app -> update state -> commit block to app -> update store height and state hash.
 // As the entire process can't be atomic we need to make sure the following condition apply before
-// we're applying the block in the happy path: block height - 1 == abci app last block height.
-// In case the following doesn't hold true, it means we crashed after the commit and before updating the store height.
-// In that case we'll want to align the store with the app state and continue to the next block.
+// - block height is the expected block height on the store (height + 1).
+// - block height is the expected block height on the app (last block height + 1).
 func (m *Manager) applyBlock(ctx context.Context, block *types.Block, commit *types.Commit, blockMetaData blockMetaData) error {
-	if block.Header.Height != m.store.Height()+1 {
-		// We crashed after the commit and before updating the store height.
-		m.logger.Error("Block not applied. Wrong height", "block height", block.Header.Height, "store height", m.store.Height())
-		return nil
+	// TODO (#330): allow genesis block with height > 0 to be applied.
+	// TODO: add switch case to have defined behavior for each case.
+	//validate block height
+	if block.Header.Height != m.store.NextHeight() {
+		m.logger.Error("Block not applied. wrong height", "block height", block.Header.Height, "expected height", m.store.NextHeight())
+		return types.ErrInvalidBlockHeight
 	}
 
 	m.logger.Debug("Applying block", "height", block.Header.Height, "source", blockMetaData.source)
 
 	// Check if alignment is needed due to incosistencies between the store and the app.
-	isAlignRequired, err := m.alignStoreWithApp(ctx, block)
+	// In the happy path we expect block height - 1 == abci app last block height.
+	isAlignRequired, err := m.alignStoreWithAppIfNeeded(block)
 	if err != nil {
 		return err
 	}
+	// In case the following true, it means we crashed after the commit and before updating the store height.
+	// In that case we'll want to align the store with the app state and continue to the next block.
 	if isAlignRequired {
```

Comment:
isAlignRequired is a confusing name because it says something _will_ happen, but the alignment already happened in alignStoreWithApp

Reply:
`didAlign`

Reply:
either that or maybe rename `alignStoreWithAppIfNeeded` instead